### PR TITLE
WT-7833 In block manager, coordinate multiple readers/writers with flush_tier

### PR DESCRIPTION
**Pull request by ddanderson**

### CODE REVIEW

**Comment by ddanderson** [comment]
@keithbostic @sueloverso @keitharnoldsmith Can you please review?  

This solves the two issues I had been seeing - accesses to a freed WT_BLOCK and a WT_BLOCK ptr that was NULL. I'll continue to test to see if anything else comes up. @keithbostic, I don't know if this is what you had in mind for this work. I'd be happy to talk through any of this.

File: test/csuite/tiered_abort/main.c [hunk 1]
```diff
@@ -131,11 +131,6 @@ typedef struct {
     uint32_t info;
 } THREAD_DATA;
 
-/*
- * TODO: WT-7833 Lock to coordinate inserts and flush_tier. This lock should be removed when that
```
Comment (sueloverso) [inline test/csuite/tiered_abort/main.c:135]: There are other csuite tests with this identical usage (possibly wtperf also). They should be fixed. test/schema_abort, test/random_directio. I had opened a ticket to clean those up after in WT-9425. But feel free to do it here and mark that a duplicate.

**Review by sueloverso — APPROVED** [review]
I don't know if we want this merged as it serializes pretty much all of tiered storage. But this PR also illustrates and identifies the areas needing something. So LGTM in general with a few minor nits. But there should be a larger discussion about the general direction.

**Comment by ddanderson** [comment]
@sueloverso,
> it serializes pretty much all of tiered storage

I don't think that's the case.  Most BM operations briefly hold a read lock, so they can safely traverse `bm->block` without it changing under them.  Once they have a block, the read lock is released and they can do their operation `__bm_read`, `__bm_write`, whatever.  The exceptions are` __bm_switch_object`, and `__bm_close`.  Switch_object holds the write lock while it creates a new `bm->block`.  And close holds it while it decides if the block can be freed.

@keithbostic has pointed out that all the locking will be awful for the hw caches, esp during checkpoint when there's lots of BM activity.  (There might be some ways to mitigate this).  But I don't think we're serializing much here.

**Comment by sueloverso** [comment]
@ddanderson thanks for the explanation. LGTM then and we can find any pain points and address them.

**Review by keitharnoldsmith — COMMENTED** [review]
This deserves more than a cursory review, and unfortunately I don't have time for that right now as I'm about to drop offline for a few days (at least long enough to fly back to Boston!) Hope you don't mind a bit of a delay for my feedback.

**Comment by ddanderson** [comment]
@keitharnoldsmith Don't mind a delay at all. I'll continue doing testing, and we can finish the review next week.

**Review by keitharnoldsmith — COMMENTED** [review]
I have a high-level question about this approach.

In WT-8836, we decided to avoid the overhead (i.e., at the hardware caches) of reference counting outstanding writes on a block. Instead we wait for a checkpoint to be certain that all of the writes to block have completed before we flush it to the cloud. 

In contrast, in this PR `block->active` is basically doing that reference counting. I don't know if there is a viable approach here that would avoid that overhead. So maybe this is a necessary cost. 

I don't really have an opinion right now about the best approach.  We don't have data quantifying the reference counting and locking overhead. It might turn out to be small compared to the cost of trapping into the OS for a read or write system call.  Alternatively, we've learned a bit since implementing WT-8836 about the awkwardness of requiring a checkpoint after every flush, so there are downsides with that approach, as well.

My question here is whether it makes sense to use two different techniques to figure out when IO (or writes) have drained from a block.

**Comment by agorrod** [comment]
@ddanderson I'm trying to close out older PRs. Can this be closed now, since the alternative implementation seems to be preferred?

**Comment by agorrod** [comment]
ping @ddanderson

**Comment by ddanderson** [comment]
@agorrod Yes, I'll close it.  And sorry I missed your first ping!

**Pull request closed without merging**